File: tree/mirror Tree.py

class Node:

    def __init__(self, data) -> None:
        self.data = data
        self.left = None
        self.right = None

def mirriorTree(node):
    # print(node.data, end=' ')
    if(not node):
        return
    mirriorTree(node.left)
    mirriorTree(node.right)
    swap = node.right
    node.right = node.left
    node.left = swap
    # print(node.right.data if node.right else '', node.left.data if node.left else '')
 
 
# Function to convert a given binary tree into its mirror
def convertToMirror(root):
 
    # base case: if the tree is empty
    if root is None:
        return
 
    # convert left subtree
    convertToMirror(root.left)
 
    # convert right subtree
    convertToMirror(root.right)
 
    # swap left subtree with right subtree
    temp = root.left
    root.left = root.right
    root.right = temp

def preorder(root):
 
    if root is None:
        return
 
    print(root.data, end=' ')
    preorder(root.left)
    preorder(root.right)

tree = Node(5)
tree.left = Node(6)
tree.right = Node(7)
tree.left.left = Node(2)
tree.left.right = Node(4)

preorder(tree)
print()
mirriorTree(tree)
# convertToMirror(tree)

preorder(tree)

File: dsaC++/trees/mirrorTree.cpp
#include<iostream>
#include <queue>

using namespace std;

class Node{
    public:
        int data;
        Node* left;
        Node* right;
    Node(int data){
        this->data = data;
        left = NULL;
        right = NULL;
    }
};

void InorderTraversal(Node* head){
    if(!head){
        return;
    }
    InorderTraversal(head->left);
    cout <<head->data <<" ";
    InorderTraversal(head->right);
}

void mirrorTree(Node* head, Node* &newTree){
    if(!head){
        return;
    }
    newTree = new Node(head->data);
    mirrorTree(head->left, newTree->right);
    mirrorTree(head->right, newTree->left);
}


int main(){
    Node* head = new Node(10);
    head->left = new Node(20);
    head->right = new Node(30);
    head->right->right = new Node(40);
    head->left->left = new Node(280);

    cout <<endl;
    InorderTraversal(head);
    Node* mirror = NULL;
    mirrorTree(head, mirror);
    cout <<endl;
    InorderTraversal(mirror);    
}